src/ondil/distributions/mv_normal_low_rank.py

```diff
@@ -0,0 +1,362 @@
+from itertools import product
+from typing import Dict
+
+import numpy as np
+import scipy.stats as st
+
+from ..base import Distribution, LinkFunction, MultivariateDistributionMixin
+from ..links import Identity, Log, MatrixDiag
+from ..types import ParameterShapes
+
+
+def batched_log_lilkelihood_normal_precision_low_rank(y, mu, mat_d, mat_v):
+    """Fast evaluation of the batched log likelihood."""
+    # k = y.shape[1]
+    # cov = np.linalg.inv(mat_d + mat_v @ np.swapaxes(mat_v, -2, -1))
+    # part1 = - k/2 * np.log(2 * np.pi)
+    # part2 = - 1/2 * np.log(np.linalg.det(cov))
+    # part3 = - 1 / 2 * np.sum((y - mu) * (np.linalg.inv(cov) @ (y - mu)[..., None]).squeeze(), 1)
+    # return part1 + part2 + part3
+    k = y.shape[1]
+    precision = mat_d + mat_v @ np.swapaxes(mat_v, -2, -1)
+    part1 = -k / 2 * np.log(2 * np.pi)
+    part2 = 1 / 2 * np.log(np.linalg.det(precision))
+    part3 = -1 / 2 * np.sum((y - mu) * (precision @ (y - mu)[..., None]).squeeze(), 1)
+    return part1 + part2 + part3
+
+
+def element_cov(mat_d, mat_v, i, j):
+    unit = np.expand_dims(np.eye(mat_v.shape[-1]), 0)
+    inner = np.swapaxes(mat_v, -1, -2) @ np.linalg.inv(mat_d) @ mat_v
+    inner_inverse = np.linalg.inv(unit + inner)
+    element = int(i == j) * 1 / mat_d[:, i, i] - np.sum(
+        inner_inverse
+        * (
+            (mat_v[:, i, :] / np.expand_dims(mat_d[:, i, i], -1))[:, :, None]
+            @ (mat_v[:, j, :] / np.expand_dims(mat_d[:, j, j], -1))[:, None, :]
+        ),
+        axis=(-2, -1),
+    )
+    return element
+
+
+def partial1_mu_element(y, mat_mu, mat_d, mat_v, i):
+    term1 = mat_d[:, i, i] * (y - mat_mu)[:, i]
+    term2 = np.sum(
+        mat_v[:, [i], :] * mat_v * np.expand_dims(y - mat_mu, -1), axis=(-2, -1)
+    )
+    # This is a slightly cleaner version of below
+    # term2 = np.squeeze(mat_v[:, [i], :] @ mat_v.swapaxes(-1, -2) @ np.expand_dims((y - mat_mu), -1))
+    return term1 + term2
+
+
+def partial2_mu_element(y, mat_mu, mat_d, mat_v, i):
+    # Diagonal elements of the inverse covariance matrix
+    return -(mat_d[:, i, i] + np.sum(mat_v[:, i] ** 2, 1))
+
+
+def partial1_D_element(y, mat_mu, mat_d, mat_v, i):
```

Comment:
This pattern repeats throughout the codebase. If a function is only used once, its hard to justify to even define it as a function I guess.